src/azul/chalice.py

```diff
@@ -504,34 +497,120 @@ def swagger_resource(self, file) -> Response:
+
+        def __call__(self, f):
             assert isinstance(f, chalice.app.EventSourceHandler), f
-            f.errors_threshold = errors
-            f.throttles_threshold = throttles
+            try:
+                metric_alarms = getattr(f, 'metric_alarms')
+            except AttributeError:
+                metric_alarms = f.metric_alarms = []
+            metric_alarms.append(self)
             return f
 
-        return wrapper
+        def bind(self, lambda_name: str, handler_name: Optional[str] = None, /) -> Self:
+            return attr.evolve(self,
+                               lambda_name=lambda_name,
+                               handler_name=handler_name)
+
+        @property
+        def lambda_tf_resource_name(self) -> str:
+            if self.handler_name is None:
+                return self.lambda_name
+            else:
+                assert self.handler_name != ''
+                return f'{self.lambda_name}_{self.handler_name}'
+
+        @property
+        def alarm_tf_resource_name(self) -> str:
+            return f'{self.lambda_tf_resource_name}_{self.metric.name}'
+
+    # noinspection PyPep8Naming
+    @attr.define(frozen=True)
+    class retry:
+        """
+        Use this decorator to specify the number of times a Lambda invocation of
+        the decorated event handler function should be retried. This decorator
+        cannot be used to decorate view functions, i.e. functions also decorated
+        with ``@app.route``.
+
+        https://docs.aws.amazon.com/lambda/latest/dg/invocation-retries.html
+        """
+        times: int
+        app_name: Optional[str] = attr.ib(default=None)
+        handler_name: Optional[str] = attr.ib(default=None)
+
+        def __call__(self, f):
+            assert isinstance(f, chalice.app.EventSourceHandler), f
+            f.retry = self.times
+            return f
+
+        @property
+        def tf_resource_name(self) -> str:
+            assert all((self.handler_name, self.app_name)), 'No handler or app name given'
+            app_name, _ = config.unqualified_resource_name(self.app_name)
+            return f'{app_name}_{self.handler_name}'
+
+    def retries(self) -> list[retry]:
+        retries = []
+        for handler_name, handler in self.handler_map.items():
+            if isinstance(handler, chalice.app.EventSourceHandler):
+                try:
+                    retry = getattr(handler, 'retry')
+                except AttributeError:
+                    # This is acceptable since only a selected few handlers of
+                    # type `EventSourceHandler` explicitly define a retry value
+                    # for dealing with failed Lambda executions.
+                    pass
+                else:
+                    retries.append(self.retry(times=retry,
+                                              app_name=self.app_name,
+                                              handler_name=handler_name))
```

Comment:
Nitpick: bodies of these methods would be slightly shorter, clearer, and more efficient if they were changed to generators

```suggestion
                    yield self.retry(times=retry,
                                              app_name=self.app_name,
                                              handler_name=handler_name)
```